content/post/104-dht-deep-dive.md

```diff
@@ -0,0 +1,154 @@
+
+## Two IPFS DHTs: Public and Local
+
+While many of our users utilize the publicly shared DHT to discover and advertise content, some of them operate in segregated networks (e.g., local networks or isolated VPNs). For these users, having a DHT where all non-publicly dialable nodes are clients is very problematic since NONE of the nodes are publicly dialable.
+
+To make things easier for this use case, we have _added a second DHT_ that is meant to include nodes that are not part of the public network, such as VPNs, CJDNS, Yggdrasil, etc. For now, we’ll refer to this as the LAN DHT, as opposed to the public network which is the WAN DHT. These two DHTs are separated by utilizing different DHT protocol names (i.e., `/ipfs/kad/1.0.0` for the WAN DHT, and `/ipfs/lan/kad/1.0.0` for the LAN DHT) to eliminate any accidental merging of the two networks. However, all of the non-public networks do have some risk of merging if users have not properly configured their networks.
+
+The main implementation difference between the WAN and LAN DHTs are the acceptance criteria for peers—which are eligible to be part of a routing table or query and which are not. The WAN DHT’s criteria is “do you look like a public address” and the LAN DHT’s criteria is “do you look like a non-public address.” However, while WAN DHT nodes switch from client to server mode based on whether they are publicly dialable, nodes are always LAN DHT servers (unless the “dhtclient” option has been set).
+
+## Routing Tables
+
+As mentioned earlier, every peer in a Kademlia network maintains links to various other peers in the network. The way this works is as follows:
+
+1. When we connect to a peer, check if it qualifies to be added to our routing table
+    - Ensure the peer is a DHT server that is advertising the DHT protocol ID (in IPFS `/ipfs/kad/1.0.0` for the public/WAN DHT and `/ipfs/lan/kad/1.0.0` for the LAN DHT)
+    - Ensure the peer has IP addresses that match the ranges we expect (e.g. members of the public DHT having at least one public range IP address as opposed to only addresses like `192.168.X.Y`)
+2. If it qualifies, then determine how close the new peer is in the Kademlia address space to us to figure out which “bucket” it should go into (i.e. if the peer is between 2^7 and 2^8 away from us and the address space is of size 2^256 then the peer goes into bucket 256-8)
+3. Attempt to put the peer in the bucket
+    - If the bucket is not full (i.e., has less than 20 peers in it) then add the peer
+    - If the bucket is full then determine if there are any peers that have not been “recently useful” to us, and then drop one of them and replace it with the new peer (“recently” and “useful” are defined below). Otherwise, don’t add the peer to the bucket
```

Comment:
We've somewhat tossed/modified this concept since the release of v0.5. @petar what do you think about changing this to:

```suggestion
    - If the bucket is full then determine if there are any peers that are "replaceable" (defined below), and then drop one of them and replace it with the new peer. Otherwise, don’t add the peer to the bucket
```